tm2/pkg/bft/privval/signer/local/key.go

```diff
@@ -0,0 +1,142 @@
+
+	filePath string
+}
+
+// FileKey validation errors.
+var (
+	errInvalidPrivateKey = errors.New("invalid private key")
+	errPublicKeyMismatch = errors.New("public key does not match private key derivation")
+	errAddressMismatch   = errors.New("address does not match public key")
+	errFilePathNotSet    = errors.New("filePath not set")
+)
+
+// validate validates the FileKey.
+func (fk *FileKey) validate() error {
+	// Make sure the private key is set.
+	if fk.PrivKey == nil {
+		return errInvalidPrivateKey
+	}
+
+	// Make sure the public key is derived from the private one.
+	if !fk.PrivKey.PubKey().Equals(fk.PubKey) {
+		return errPublicKeyMismatch
+	}
+
+	// Make sure the address is derived from the public key.
+	if fk.PubKey.Address().Compare(fk.Address) != 0 {
+		return errAddressMismatch
+	}
+
+	// Check if the file path is set.
+	if fk.filePath == "" {
+		return errFilePathNotSet
+	}
+
+	return nil
+}
+
+// save persists the FileKey to its file path.
+func (fk *FileKey) save() error {
+	// Check if the FileKey is valid.
+	if err := fk.validate(); err != nil {
+		return err
+	}
+
+	// Marshal the FileKey to JSON bytes using amino.
+	jsonBytes := amino.MustMarshalJSONIndent(fk, "", "  ")
+
+	// Write the JSON bytes to the file.
+	if err := osm.WriteFileAtomic(fk.filePath, jsonBytes, 0o600); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// LoadFileKey reads a FileKey from the given file path.
+func LoadFileKey(filePath string) (*FileKey, error) {
+	// Read the JSON bytes from the file.
+	rawJSONBytes, err := os.ReadFile(filePath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the JSON bytes into a FileKey using amino.
+	fk := &FileKey{}
+	err = amino.UnmarshalJSON(rawJSONBytes, fk)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshal FileKey from %v: %w", filePath, err)
+	}
+
+	// Manually set the private file path.
+	fk.filePath = filePath
+
+	// Validate the FileKey.
+	if err := fk.validate(); err != nil {
+		return nil, err
+	}
+
+	return fk, nil
+}
+
+// GenerateFileKey generates a new random FileKey.
+func GenerateFileKey() *FileKey {
+	// Generate a new random private key.
+	privKey := ed25519.GenPrivKey()
+
+	// Create a new FileKey instance.
+	return &FileKey{
+		PrivKey: privKey,
+		PubKey:  privKey.PubKey(),
+		Address: privKey.PubKey().Address(),
+	}
+}
+
+// GeneratePersistedFileKey generates a new random FileKey persisted to disk.
+func GeneratePersistedFileKey(filePath string) (*FileKey, error) {
+	// Generate a new random FileKey.
+	fk := GenerateFileKey()
+
+	// Set the file path.
+	fk.filePath = filePath
+
+	// Persist the FileKey to disk.
+	if err := fk.save(); err != nil {
+		return nil, err
+	}
+
+	return fk, nil
+}
+
+// NewFileKey returns a new FileKey instance from the given file path.
+// If the file does not exist, a new FileKey is generated and persisted to disk.
+func NewFileKey(filePath string) (*FileKey, error) {
```

Comment:
This can be a bit misleading, the func is called `NewFileKey` and gives a path. I expect that when I call this, it's gonna error in case the key is not on the given path.

Instead, it creates a fresh state on that path and returns it

Reply:
It doesn't seems weird to me that a function starting with ⁠`New` creates something and I mentioned this behaviour in the associated comment. 

But if you find it misleading, I can rename it. What do you suggest? ⁠`LoadOrCreateFileKey`?

Reply:
Let's go with `LoadOrMakeFileKey` 🥲 

![9ojtzf](https://github.com/user-attachments/assets/d5069cf2-3b7f-4f4a-80b0-4b6ba656253d)

Reply:
Done https://github.com/gnolang/gno/pull/3554/commits/c43401f7f4600cf495fab89c4de9b323b741aeeb